runtime/src/network/audited.rs

```diff
@@ -0,0 +1,176 @@
+use crate::{deterministic::Auditor, Error, SinkOf, StreamOf};
+use sha2::Digest;
+use std::{net::SocketAddr, sync::Arc};
+
+/// A sink that audits network operations.
+pub struct Sink<S: crate::Sink> {
+    auditor: Arc<Auditor>,
+    inner: S,
+    remote_addr: SocketAddr,
+}
+
+impl<S: crate::Sink> crate::Sink for Sink<S> {
+    async fn send(&mut self, data: &[u8]) -> Result<(), Error> {
+        self.auditor.event(b"send_attempt", |hasher| {
+            hasher.update(self.remote_addr.to_string().as_bytes());
+            hasher.update(data);
+        });
+
+        self.inner.send(data).await.inspect_err(|e| {
+            self.auditor.event(b"send_failure", |hasher| {
+                hasher.update(self.remote_addr.to_string().as_bytes());
+                hasher.update(e.to_string().as_bytes());
+            });
+        })?;
+
+        self.auditor.event(b"send_success", |hasher| {
+            hasher.update(self.remote_addr.to_string().as_bytes());
+            hasher.update(data);
```

Comment:
I chose what I thought would be reasonable fields to include in the audit hash. Lmk if you want me to change any of them. Here, for example, we arguably don't need to include `data` in the hash.